### PR TITLE
Transactional and Queued write modes

### PR DESCRIPTION
**Pull request by ohadbitt**
New write mode - queued mode which doesn't create a temporary table and does not poll on ingestion result

### CODE REVIEW

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 1]
```diff
@@ -94,83 +94,115 @@ object KustoWriter {
...
+              writeOptions, crp, tableExists, rdd.sparkContext, authentication)
+          }
+          asyncWork.onFailure {
+            case exception: Exception =>
+              if (writeOptions.userTempTableName.isEmpty) {
+                kustoClient.cleanupIngestionByProducts(
+                  tableCoordinates.database, kustoClient.engineClient, tmpTableName, crp)
+              }
+              KDSU.reportExceptionAndThrow(myName, exception, "writing data", tableCoordinates.clusterUrl, tableCoordinates.database, table, shouldNotThrow = true)
+              KDSU.logError(myName, "The exception is not visible in the driver since we're in async mode")
```
Comment (yihezkel) [inline connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:140]: I know it was already like this, but do we want a separate error trace for this? I'd think it'd be better to consolidate both lines.

File: connector/src/test/scala/com/microsoft/kusto/spark/KustoSourceE2E.scala [hunk 2]
```diff
@@ -60,42 +60,42 @@ class KustoSourceE2E extends FlatSpec with BeforeAndAfterAll {
...
-        KustoSourceOptions.KUSTO_TRANSIENT_STORAGE -> storage.toString,
-      KustoSourceOptions.KUSTO_USER_PROMPT -> true.toString
-      //      KustoSourceOptions.KUSTO_BLOB_STORAGE_ACCOUNT_NAME -> storageAccount,
-      //      KustoSourceOptions.KUSTO_BLOB_STORAGE_ACCOUNT_KEY -> blobKey,
-      //      KustoSourceOptions.KUSTO_BLOB_CONTAINER -> container
-    )
-
-    spark.read.kusto(cluster, database, query, conf).show(20)
-  }
+//
```
Comment (yihezkel) [inline connector/src/test/scala/com/microsoft/kusto/spark/KustoSourceE2E.scala:63]: Why not `/* */`?

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoSinkOptions.scala [hunk 3]
```diff
@@ -73,5 +94,7 @@ case class WriteOptions(pollingOnDriver:Boolean = true,
                         autoCleanupTime: FiniteDuration,
                         maxRetriesOnMoveExtents: Int = 10,
                         minimalExtentsCountForSplitMerge: Int = 400,
-                        adjustSchema: SchemaAdjustmentMode.SchemaAdjustmentMode = SchemaAdjustmentMode.NoAdjustment)
+                        adjustSchema: SchemaAdjustmentMode.SchemaAdjustmentMode = SchemaAdjustmentMode.NoAdjustment,
+                        isTransactionalMode: Boolean = false,
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoSinkOptions.scala:98]: I will set this to true and change to false on major

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 4]
```diff
@@ -94,83 +94,115 @@ object KustoWriter {
...
       KDSU.logInfo(myName, s"$IngestSkippedTrace '$table'")
     } else {
-      // KustoWriter will create a temporary table ingesting the data to it.
-      // Only if all executors succeeded the table will be appended to the original destination table.
-      kustoClient.initializeTablesBySchema(tableCoordinates, tmpTableName, data
-        .schema, targetSchema, writeOptions, crp, stagingTableIngestionProperties.creationTime == null)
-      KDSU.logInfo(myName, s"Successfully created temporary table $tmpTableName, will be deleted after completing the operation")
+      if (writeOptions.userTempTableName.isDefined) {
+        if (kustoClient.engineClient.execute(tableCoordinates.database,
+          generateTableGetSchemaAsRowsCommand(tableCoordinates.table.get), crp).getPrimaryResults.count() <= 0 ||
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:99]: yes thanks - i just put the wrong table name here

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala [hunk 5]
```diff
@@ -94,83 +94,115 @@ object KustoWriter {
...
+          !tableExists) {
+          throw new InvalidParameterException("Temp table name provided but the table does not exist. Either drop this " +
+            "option or create the table beforehand.")
+        }
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:103]: if a user provided a temporary table and it doesnt exist i dont want to not use it behind his back - he should be aware of the false usage

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 1 above)
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:140]: I do want it in a separate trace - this should raise awareness for why couldn't the user handle the error.. this is printed at most once in a run - why does it matter?

File: connector/src/test/scala/com/microsoft/kusto/spark/KustoSourceE2E.scala (hunk 2 above)
Comment (ohadbitt) [reply connector/src/test/scala/com/microsoft/kusto/spark/KustoSourceE2E.scala:63]: This is misstake - anyway this file is rewritten in another PR

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 1 above)
Comment (yihezkel) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:138]: NBD
But if the 2nd trace is only explaining the 1st, I'd think it's confusing to separate them.

But you're saying the 2nd trace is relevant to any exception thrown during that run?

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (yihezkel) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:101]: I'm referring to adding that else **after** line 103 above (as an `else` to `if (kustoClient.engineClient.execute(tableCoordinates.database, generateTableGetSchemaAsRowsCommand(writeOptions.userTempTableName.get), crp).getPrimaryResults.count() <= 0 || !tableExists)`)

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 1 above)
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:138]: Yes

File: connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ohadbitt) [reply connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala:101]: oh 
its already done when the tempTableName is instantiated

**Review by yihezkel — APPROVED** [review]
The code looks great!

One small request (best if in this PR): I think when adding a new feature (like when separating Transactional vs Queued modes), we should include 1 E2E test. And regardless, I think UTs are almost always important.

**Pull request merged**